pkg/networkmanager/firewall-client.js

```diff
@@ -277,36 +292,56 @@ firewall.createService = (service, name, ports) => {
                                ]);
 };
 
-firewall.enableService = (service) => {
-    return getDefaultZonePath()
-            .then(path => firewalld_dbus.call(path, 'org.fedoraproject.FirewallD1.config.zone',
-                                              'addService', [service]));
+firewall.enableService = (zones, service) => {
+    // We can't use Promise.all() here until cockpit is able to dispatch es2015 promises
+    // https://github.com/cockpit-project/cockpit/issues/10956
+    // eslint-disable-next-line cockpit/no-cockpit-all
+    return cockpit.all(
+        zones.map(z => firewalld_dbus.call('/org/fedoraproject/FirewallD1/config',
+                                           'org.fedoraproject.FirewallD1.config',
+                                           'getZoneByName', [z])
+                .then(path => firewalld_dbus.call(path[0], 'org.fedoraproject.FirewallD1.config.zone',
+                                                  'addService', [service]))
+                .then(() => getZones())
+                .catch(error => console.warn(error))));
 };
 
 /*
- * Add a predefined firewalld service to the default zone (i.e., open its
+ * Add a predefined firewalld service to the specified zone (i.e., open its
  * ports).
  *
  * Returns a promise that resolves when the service is added.
  */
-firewall.addService = (service) => {
+firewall.addService = (zone, service) => {
     return firewalld_dbus.call('/org/fedoraproject/FirewallD1',
                                'org.fedoraproject.FirewallD1.zone',
-                               'addService', ['', service, 0])
-            .then(reply => getDefaultZonePath())
-            .then(path => firewalld_dbus.call(path, 'org.fedoraproject.FirewallD1.config.zone',
+                               'addService', [zone, service, 0])
+            .then(reply => firewalld_dbus.call('/org/fedoraproject/FirewallD1/config',
+                                               'org.fedoraproject.FirewallD1.config',
+                                               'getZoneByName', [zone]))
+            .then(path => firewalld_dbus.call(path[0], 'org.fedoraproject.FirewallD1.config.zone',
                                               'addService', [service]))
```

Comment:
Does that also cause the service to be added to the runtime config, or  just the persistent one? I think it should also do runtime, for consistency. Can you please explain the duality and strategy wrt. runtime vs. persistent in the commit message?

Reply:
Both weirdly enough, if you add a custom service to the runtime configuration it'll cause an error.

I'll update the commit message.

Reply:
So by that logic, why doesn't this get enabled in the runtime config here, and only in the persistent config?

Reply:
I must've implemented this badly when I tried it, indeed you can add it to the runtime config and then to the persistent one, although somehow it seems to be active right away regardless.

Anyway let's be consistent and add it to runtime, then permanent.

Reply:
That sounds like a bug -- it would mean that adding a service to a zone isn't immediately effective? If it is, then this  should be validated with a CLI test (`iptables -L`, or at least with `firewall-cmd`)